File: leetcode/0001.two-sum/two-sum.js
/**
 * 暴力解: O(n^2)
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
 var twoSumSol1 = function(nums, target) {
    for (let i = 0; i < nums.length-1; i++) {
        for (let j = i+1; j < nums.length; j++) {
            if (nums[i]+nums[j] === target) {
                return [i, j]
            }
        }
    }

    return null
 }

/**
 * Map解: O(n)
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSumSol2 = function(nums, target) {
    const numMap = new Map()

    for (let i = 0; i < nums.length; i++) {
        if (numMap.has(target-nums[i])) {
            return [i, numMap.get(target-nums[i])]
        }
        numMap.set(nums[i], i)
    }

    return null
 }

File: leetcode/0001.two-sum/two-sum.go
package leetcode

// 暴力解: O(n^2)
func twoSumSol1(nums []int, target int) []int {
	var result = make([]int, 2)

	for i := 0; i < len(nums)-1; i++ {
		for j := i + 1; j < len(nums); j++ {
			if nums[i]+nums[j] == target {
				result[0] = i
				result[1] = j
			}
		}
	}

	return result
}

// Map解: O(n)
func twoSumSol2(nums []int, target int) []int {
	numMap := make(map[int]int)

	for i := 0; i < len(nums); i++ {
		if v, ok := numMap[target-nums[i]]; ok {
			return []int{i, v}
		}

		numMap[nums[i]] = i
	}

	return nil
}
